pkg/terraformer/terraformer.go

```diff
@@ -205,3 +221,65 @@ func (t *Terraformer) executeTerraform(ctx context.Context, command Command) err
 	log.Info("terraform process finished successfully", "command", command)
 	return nil
 }
+
+func (t *Terraformer) addFinalizer(ctx context.Context) error {
+	logger := t.log.WithName("add-finalizer")
+	return t.updateObjectFinalizers(ctx, logger, controllerutil.AddFinalizer)
+
+}
+
+func (t *Terraformer) removeFinalizer(ctx context.Context) error {
+	logger := t.log.WithName("remove-finalizer")
+	return t.updateObjectFinalizers(ctx, logger, controllerutil.RemoveFinalizer)
+}
+
+func (t *Terraformer) updateObjectFinalizers(ctx context.Context, log logr.Logger, patchObj func(obj controllerutil.Object, finalizerName string)) error {
+	objects := []controllerutil.Object{
+		&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: t.config.Namespace,
+				Name:      t.config.VariablesSecretName,
+			},
+		},
+		&corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: t.config.Namespace,
+				Name:      t.config.ConfigurationConfigMapName,
+			},
+		},
+		&corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: t.config.Namespace,
+				Name:      t.config.StateConfigMapName,
+			},
+		},
+	}
+	log.V(3).Info("updating finalizers for terraform resources", "namespace", t.config.Namespace)
+
+	for _, obj := range objects {
+		key, err := client.ObjectKeyFromObject(obj)
+		if err != nil {
+			log.Error(err, "failed to construct key", "key", obj, "error")
+			return err
+		}
+
+		err = t.client.Get(ctx, key, obj)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.V(1).Info("skipping non-existing object", "key", obj)
+				continue
+			}
+			log.Error(err, "failed to get object", "key", obj, "error")
+			return err
+		}
+
+		old := obj.DeepCopyObject()
+		patchObj(obj, TerraformerFinalizer)
+		if err := t.client.Patch(ctx, obj, client.MergeFrom(old)); client.IgnoreNotFound(err) != nil {
```

Comment:
Can you use optimistic locking (also send the resourceVersion) and retry on conflict here?
I think, json merge patch replaces the whole array instead of operating on single items, so it is not safe for concurrent use and requires locking.
```suggestion
		if err := t.client.Patch(ctx, obj, client.MergeFromWithOptions(old, client.MergeFromWithOptimisticLock{})); client.IgnoreNotFound(err) != nil {
```

Reply:
As long as we are not touching the `.data` field in the patch, it should not change anything there.
I have added the Optimistic Lock and also make it retry on conflict.